---
title: "Testing annotation on the alignment"
output: null_document
---

# Setup
```{r}
library(ggplot2)
library(cowplot)
library(rhdf5)

path.pannagram = '../'  # the folder with the cloned repo

source(paste(path.pannagram, 'utils/utils.R', sep = ''))
source(paste(path.pannagram, 'visualisation/visualisation.R', sep = ''))  # plotPanAcc
source(paste(path.pannagram, 'visualisation/panplot.R', sep = ''))  # ggSynteny
source(paste(path.pannagram, 'pangen/comb_func.R', sep = ''))  # getPrevNext
source(paste(path.pannagram, 'analysis/analys_func.R', sep = ''))  # gffgff

```

## Paths-Folders
```{r}
path.work = '../../work/004_test_pan/data_greg/'  # change to yours

path.annot = paste(path.work, 'annotation/', sep = '')
path.cons = paste(path.work, 'consensus/', sep = '')  # the result of pannagram
path.out = paste(path.work, 'out/', sep = '')
path.figures = paste(path.work, 'figures/', sep = '')

if (!dir.exists(path.out))  dir.create(path.out)  # Check if the directory exists
if (!dir.exists(path.figures))  dir.create(path.figures)  # Check if the directory exists

n.chr = 5
ref.acc = '0'

```

## Additional variables
```{r}

gr.accs.e <- "accs/"
gr.accs.b <- "/accs"

```

## Colors
```{r}

c.red = '#872341'
c.grey = '#EBEBEB'

```


# Show pangenome coordinates
## Pairwise alignments
```{r}
# Set chromosome index and setup the corresponding MSA file
i.chr = 4
file.msa = paste(path.cons, 'msa_', i.chr, '_', i.chr, '_ref_',ref.acc,'.h5', sep = '')

# Read group information from the MSA file and extract accession names and pangenome length
groups = h5ls(file.msa)
accessions = groups$name[groups$group == gr.accs.b]
pan.len = unique(groups$dim[groups$group == gr.accs.b])
if(length(pan.len) != 1) stop('Something is wrong in the alignment')

# Create plots for each accession
p.list = list()
for(acc in accessions){
  # Create ggplot object for corresponcence between pangenome coordinate and an accession
  p.list[[acc]] = plotPanAcc(file.msa, acc)
}

# Combine all plots and display
p.combo <- plot_grid(plotlist = p.list, ncol = 3)
p.combo

# Save the combined plot to a PDF file
pdf(paste(path.figures, 'pairwise_pan_acc_chr_',i.chr,'.pdf', sep = ''), width = 8, height = 4)
print(p.combo)
dev.off()


```

## Common alignment plot (grey shades)
```{r}

# Load MSA blocks information and set window size
file.blocks = paste(path.cons, 'msa_blocks_ref_', ref.acc, '.rds', sep = '')
wnd.size = 100000

# Read block indices and prepare plots for each of the first 5 chromosomes
idx.blocks = readRDS(file.blocks)
p.list = list()
for(i.chr in 1:5){
  p.list[[i.chr]] = ggSynteny(idx.blocks, i.chr, unique(idx.blocks$acc), c(1, 3, 2), wnd.size=wnd.size)
}

# Combine and display all chromosome plots
p.combo <- plot_grid(plotlist = p.list, nrow = 5)
p.combo

# Save the combined plot to a PDF file
pdf(paste(path.figures, 'msa_pangenone', '.pdf', sep = ''), width = 8, height = 6)
print(p.combo)
dev.off()


```

# Table of correspondence
```{r}
# Define chromosome and MSA file path
i.chr = 4
file.msa = paste(path.cons, 'msa_', i.chr, '_', i.chr, '_ref_',ref.acc,'.h5', sep = '')

h5ls(file.msa)

# Initialize vector and load MSA data for each accession
v = c()
for(acc in accessions){
  v.acc = h5read(file.msa, paste(gr.accs.e, acc, sep = ''))
  v = cbind(v, v.acc)
}
colnames(v) = accessions

# Show the correspondence
# v[10000000 + (-5:50),]
v[2500000 + (-5:50),]


```


## Get a GFF file for one accession using a GFF file from another accession
```{r}
# Display the list of accessions
pokaz(accessions)

# Define accessions for comparison
acc1 = 'acc_0'
acc2 = 'acc_6046-v1.1'

# Read GFF
file.gff.acc1 = paste(path.annot, 'TAIR10_GFF3_genes_transposons.gff', sep = '')
gff1 = read.table(file.gff.acc1, stringsAsFactors = F)

# Chromosome IDs
gff1$chr = as.numeric(gsub('Chr', '', gff1$V1))
gff1 = gff1[!is.na(gff1$chr),]  # remove wrong names or ChrC, ChrM, etc....

# Filtration 
gff1 = gff1[!(gff1$V3 %in% c('chromosome')),]  # unnecessary attributes
gff1 = gff1[gff1$V3 %in% c('gene', 'exon'),]  # important attributes

# Convert GFF Annotations from acc1 to acc2
gff2.res = gff2gff(path.cons, acc1, acc2, ref.acc, gff1, flag.exact = T)

# Results
gff2.remain = gff2.res$gff2.remain
gff2.loosing = gff2.res$gff2.loosing
pokaz('Annotated before:', nrow(gff1), 'Annotated now', nrow(gff2.remain), 'Lost', nrow(gff1) - nrow(gff2.remain))

# TODO: Check that all exons have genes....

# Save
write.table(gff2.remain[,1:9], paste(path.annot, acc2, '_based_on_', acc1, '.txt', sep = ''),
            quote = F, col.names = F, row.names = F, sep = '\t')


```


## Get alignment from positions (beg-end) in coordinates of accession Z
```{r}
# Define accession identifiers
acc1 = 'acc_0'
acc2 = 'acc_6046-v1.1'

# Set the chromosome number for analysis
i.chr = 1

# Define the start and end positions for the region of interest
pos.beg = 10000000
pos.end = pos.beg + 500

# Read multiple sequence alignment (MSA) data for both accessions from the HDF5 file
file.msa = paste(path.cons, 'msa_', i.chr, '_', i.chr, '_ref_',ref.acc,'.h5', sep = '')
v = cbind(h5read(file.msa, paste(gr.accs.e, acc1, sep = '')),
          h5read(file.msa, paste(gr.accs.e, acc2, sep = '')))

# Filter the MSA data to include only positions within the specified range
idx.pos = (v[,1] >= pos.beg) & (v[,1] <= pos.end)
v.corresp = v[idx.pos, 2]


```


